File: stockify_app_flutter/lib/feature/item/service/item_service_implementation.dart
import 'package:stockify_app_flutter/feature/item/model/item.dart';
import 'package:stockify_app_flutter/feature/item/model/item_filter_param.dart';
import 'package:stockify_app_flutter/feature/item/service/item_service.dart';

import '../repository/item_repository.dart';

class ItemServiceImplementation implements ItemService {
  ItemServiceImplementation._privateConstructor();

  static final ItemServiceImplementation _instance =
      ItemServiceImplementation._privateConstructor();

  static ItemService get instance => _instance;

  final ItemRepository _itemRepository = ItemRepository.instance;

  @override
  void addItem(Item item) {
    _itemRepository.addItem(item);
  }

  @override
  void deleteItem(int id) {
    _itemRepository.deleteItem(id);
  }

  @override
  List<Item> getAllItems() {
    return _itemRepository.getAllItems();
  }

  @override
  Item getItem(int id) {
    return _itemRepository.getItemById(id);
  }

  @override
  void updateItem(Item item) {
    _itemRepository.updateItem(item);
  }

  @override
  List<Item> getFilteredItems(ItemFilterParams params) {
    return _itemRepository.getFilteredItems(params);
  }
}

File: stockify_backend_golang/src/feature/item/service/item_service_implementation.go
package service

import (
	"stockify_backend_golang/src/feature/item/model"
	"stockify_backend_golang/src/feature/item/repository"
)

type itemService struct {
	repo repository.ItemRepository
}

func ItemServiceImplementation(repo repository.ItemRepository) ItemService {
	return &itemService{repo: repo}
}

func (s *itemService) AddItem(item model.Item) {
	s.repo.AddItem(item)
}

func (s *itemService) GetAllItems() []model.Item {
	return s.repo.GetAllItems()
}

func (s *itemService) GetItemById(id uint64) model.Item {
	return s.repo.GetItemById(id)
}

func (s *itemService) UpdateItem(item model.Item) {
	s.repo.UpdateItem(item)
}

func (s *itemService) DeleteItemById(id uint64) {
	s.repo.DeleteItemById(id)
}

func (s *itemService) GetFilteredItems(params model.ItemFilterParams) ([]model.Item, error) {
	return s.repo.GetFilteredItems(params)
}
